### ISSUE TITLE
Latest spurious disconnects

### ISSUE THREAD

**Issue by tomaka** [paritytech/substrate#540]
Nodes still sometimes disconnect from each other spontaneously (commit d23a89a7523ccd1a432354e0f181013e222345ba).

Here is an interesting log I have from a local testnet:

```
2018-08-10 15:09:07  TRACE libp2p_core::swarm  Swarm dialing /p2p/QmSW4RzuS9j9FUgXakXu3HZzJKiLCYCL27hEqbodAFgWba
2018-08-10 15:09:07  DEBUG libp2p_core::swarm  Error in dialer upgrade: Custom { kind: Other, error: StringError("couldn\'t find any multiaddress for peer") }
2018-08-10 15:09:07 tokio-runtime-worker-4 INFO substrate  Idle (1 peers), best: #364 (1f78…2a15)
2018-08-10 15:09:07  INFO sub-libp2p  Finishing future for proto [100, 111, 116] with PeerId(QmSW4RzuS9j9FUgXakXu3HZzJKiLCYCL27hEqbodAFgWba) => Ok(())
```

Despite the `Ok(())`, this is from the side that closed the connection.
After this, the other node detects that the substream was closed and produces an `Ok(())` as well.

It looks like failing to reaching back a node closes existing substreams.

**Comment by gavofyork** [comment]
Any progress here?

**Comment by tomaka** [comment]
Should be fixed in #742
